base/uk.ac.stfc.isis.ibex.opis/check_opi_format.py

```diff
@@ -0,0 +1,272 @@
+            err = "Error on line " + str(error.sourceline) + ": " + etree.tostring(error) \
+                  + "\n... A button was not within a grouping container."
+            self.errors.append(err)
+
+        for error in root.xpath(led_xpath):
+            err = "Error on line " + str(error.sourceline) + ": " + etree.tostring(error) \
+                  + "\n... An LED indicator was not within a grouping container."
+            self.errors.append(err)
+
+        for error in root.xpath(dropdown_xpath):
+            err = "Error on line " + str(error.sourceline) + ": " + etree.tostring(error) \
+                  + "\n... An dropdown menu was not within a grouping container."
+            self.errors.append(err)
+
+        for error in root.xpath(textinput_xpath):
+            err = "Error on line " + str(error.sourceline) + ": " + etree.tostring(error) \
+                  + "\n... A text input field was not within a grouping container."
+            self.errors.append(err)
+
+    def check_capitals_for_grouping_containers(self, root):
+        container_name_xpath = "//widget[@typeId='org.csstudio.opibuilder.widgets.groupingContainer']/name"
+
+        for name in root.xpath(container_name_xpath):
+            capitalisation_error = False
+            words = name.text.split()
+
+            # Ignore words less than 4 characters as these are probably prepositions
+            for word in words:
+                if len(word) > self.ignore_short_words_limit \
+                        and word.title() != word \
+                        and not any(s in word for s in self.ignore):
+                    capitalisation_error = True
+
+            if capitalisation_error:
+                err = "Error on line " + str(name.sourceline) + ": " + etree.tostring(name) \
+                      + "\n... Grouping container titles should be in 'Title Case'."
+                self.errors.append(err)
+
+    def check_capitals_for_labels_outside_grouping_containers(self, root):
+
+        # Select a Label outside a grouping container
+        textinput_xpath = \
+            "//widget[@typeId='org.csstudio.opibuilder.widgets.Label' " \
+            "and not(ancestor::widget[@typeId='org.csstudio.opibuilder.widgets.groupingContainer'])]/text"
+
+        for name in root.xpath(textinput_xpath):
+            capitalisation_error = False
+            words = name.text.split()
+
+            for word in words:
+                if len(word) > self.ignore_short_words_limit \
+                        and word.title() != word \
+                        and not any(s in word for s in self.ignore):
+                    capitalisation_error = True
+
+            if capitalisation_error:
+                err = "Error on line " + str(name.sourceline) + ": " + etree.tostring(name) \
+                      + "\n... Labels outside grouping containers are titles, and therefore should be 'Title Case'."
+                self.errors.append(err)
```

Comment:
The previous two methods use the same code. Put it in a separate "check_title_case" method which you call from here with xpath query and error message as arguments. Also allows you to reuse it e.g. if you decide you want to check for title case in additional elements in the future.